app/src/main/java/com/ataulm/whatsnext/WhatsNextService.java

```diff
@@ -1,75 +1,84 @@
+import com.ataulm.whatsnext.api.LetterboxdApi;
 
-import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.Callable;
 
 import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.reactivex.functions.BiFunction;
+import io.reactivex.functions.Function;
 
 public class WhatsNextService {
 
-    private final Letterboxd letterboxd;
-    private final TokensStore tokensStore;
-    private final Clock clock;
+    private final LetterboxdApi letterboxdApi;
+    private final FilmSummaryConverter filmSummaryConverter;
+    private final FilmRelationshipConverter filmRelationshipConverter;
 
-    public WhatsNextService(Letterboxd letterboxd, TokensStore tokensStore, Clock clock) {
-        this.letterboxd = letterboxd;
-        this.tokensStore = tokensStore;
-        this.clock = clock;
+    public WhatsNextService(LetterboxdApi letterboxdApi, FilmSummaryConverter filmSummaryConverter, FilmRelationshipConverter filmRelationshipConverter) {
+        this.letterboxdApi = letterboxdApi;
+        this.filmSummaryConverter = filmSummaryConverter;
+        this.filmRelationshipConverter = filmRelationshipConverter;
     }
 
     public Observable<Token> login(final String username, final String password) {
-        return Observable.fromCallable(new Callable<Token>() {
-            @Override
-            public Token call() throws Exception {
-                return letterboxd.fetchAccessToken(username, password);
-            }
-        });
+        return letterboxdApi.fetchAuthToken(username, password, "password")
+                .toObservable()
+                .map(toToken());
     }
 
-    public Observable<List<FilmSummary>> search(final String searchTerm) {
-        return Observable.fromCallable(new Callable<List<FilmSummary>>() {
+    private Function<AuthTokenApiResponse, Token> toToken() {
+        return new Function<AuthTokenApiResponse, Token>() {
             @Override
-            public List<FilmSummary> call() throws Exception {
-                return letterboxd.search(searchTerm);
+            public Token apply(AuthTokenApiResponse authTokenApiResponse) {
+                return new Token(
+                        authTokenApiResponse.getAccessToken(),
+                        authTokenApiResponse.getRefreshToken(),
+                        authTokenApiResponse.getSecondsUntilExpiry()
+                );
             }
-        });
+        };
     }
 
-    public Observable<Film> film(final String letterboxdId) {
-        return Observable.fromCallable(new Callable<Film>() {
-            @Override
-            public Film call() throws Exception {
-                return letterboxd.film(letterboxdId, getToken().getAccessToken());
-            }
-        });
-    }
-
-    private Token getToken() {
-        Token token = tokensStore.getToken();
-
-        if (token == null) {
-            throw new AuthenticationError(Type.REQUIRES_USER_SIGN_IN);
```

Comment:
This isn't handled anymore (if it was even working).

Now retrofit will return 401, but there's no error handling in place (it fails silently). I'll open an issue so I remember to handle this + the more common case where the access token will expire so we need to refresh.

Reply:
https://github.com/ataulm/whats-next/issues/11